pyat/at/errors/error_def.py

```diff
@@ -0,0 +1,449 @@
+
+        >>> ringerr = enable_errors(all=False, ShiftErr=True)
+
+        Only magnet shift errors are active
+
+        >>> ringerr = enable_errors(all=False)
+
+        No magnet errors are enabled. Only monitor errors will be active.
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`get_optics_err`
+    """
+    if not ring.has_errors:
+        ring = _apply_field_errors(ring, **kwargs)
+        ring = _apply_alignment_errors(ring, **kwargs)
+        ring._has_errors = True
+    return ring
+
+
+def has_errors(ring):
+    """Get the error status of a lattice"""
+    return getattr(ring, '_has_errors', False)
+
+
+def get_mean_std_err(ring: Lattice, key, attr, index=0):
+    vals = [np.atleast_1d(getattr(e, attr, 0.0))[index]
+            for e in ring.get_elements(key)]
+    return np.mean(vals), np.std(vals)
+
+
+def _sort_flags(kwargs):
+    errargs = {}
+    for key in _SEL_ARGS:
+        if key in kwargs:
+            errargs[key] = kwargs.pop(key)
+    return kwargs, errargs
+
+
+def find_orbit_err(ring: Lattice, refpts: Refpts = None, **kwargs):
+    """Find the closed orbit if a lattice with errors
+
+    :py:func:`find_orbit_err` enables the selected errors, finds the closed
+    orbit and modifies the result according to monitor errors.
+    The *ring* :py:class:`~.Lattice` is not modified.
+
+    Args:
+        ring:       Lattice description
+        refpts:
+
+    The *all*, *ShiftErr*, *RotationErr*, *PolynomBErr*, *PolynomBIndex*,
+    *PolynomAErr*, *PolynomAIndex* keywords are defined as in
+    :py:func:`enable_errors`
+
+    All other keywords are forwarded to :py:func:`.find_orbit`
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`.find_orbit`,
+        :py:func:`get_optics_err`, :py:func:`lattice_pass_err`
+    """
+    kwargs, errargs = _sort_flags(kwargs)
+    ring = enable_errors(ring, **errargs)
+    orbit0, orbit = find_orbit(ring, refpts=refpts, **kwargs)
+    _apply_bpm_orbit_error(ring, refpts, orbit)
+    return orbit0, orbit
+
+
+def get_optics_err(ring: Lattice, refpts: Refpts = None, **kwargs):
+    """Linear analysis of a lattice with errors
+
+    :py:func:`get_optics_err` enables the selected errors, makes the linear
+    analysis of the modified :py:class:`~.Lattice` and modifies the closed
+    orbit according to monitor errors.
+    The *ring* :py:class:`~.Lattice` is not modified.
+
+    Args:
+        ring:       Lattice description
+        refpts:
+
+    The *all*, *ShiftErr*, *RotationErr*, *PolynomBErr*, *PolynomBIndex*,
+    *PolynomAErr*, *PolynomAIndex* keywords are defined as in
+    :py:func:`enable_errors`
+
+    All other keywords are forwarded to :py:func:`.get_optics`
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`.get_optics`,
+        :py:func:`find_orbit_err`, :py:func:`lattice_pass_err`
+    """
+    kwargs, errargs = _sort_flags(kwargs)
+    ring = enable_errors(ring, **errargs)
+    ld0, bd, ld = get_optics(ring, refpts=refpts, **kwargs)
```

Comment:
ld0 closed orbit does not seem to be updated by these lines

Reply:
No, it just shows the orbit resulting from magnet errors enables in the previous line